File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.java
class Solution {
    public int minCostConnectPoints(int[][] points) {
        
        int n = points.length;
        PriorityQueue<int[]> q = new PriorityQueue<>((a,b)->a[0]-b[0]);
        boolean[] connected = new boolean[n];
        connected[0] = true;
        fill(points,q,connected,0);
        int res = 0;
        int count = 1;
        while(count<n){
            int a[] = q.poll();
            assert a != null;
            if(connected[a[1]]) continue;
            res+=a[0];
            connected[a[1]] = true;
            fill(points,q,connected,a[1]);
            count++;
        }
        return res;
      
    }
    int dist(int a[], int b[]){
        return Math.abs(a[0]-b[0]) + Math.abs(a[1]-b[1]);
    }

    void fill(int points[][], PriorityQueue<int[]> q, boolean connected[], int idx){
        int n = connected.length;
        for(int i = 0 ; i<n ; i++){
            if(!connected[i]){
                q.add( new int[]{dist(points[i],points[idx]),i});
            }
        }
    }
}

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        def manD(p1, p2):
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])

        n = len(points)
        g= {(x,y): float('inf') if i else 0 for i,(x,y) in enumerate(points)}
        ans = 0
        while g:
            x,y = min(g, key=g.get)
            ans += g.pop((x,y))
            for (i,j) in g:
                g[(i,j)] = min(g[(i,j)], manD((x,y), (i,j)))
        return ans
        